packages/transformer/src/IModelTransformer.ts

```diff
@@ -1944,6 +1944,38 @@ export class IModelTransformer extends IModelExportHandler {
       ? targetElementId
       : undefined;
 
+    if (this._options.preserveElementIdsForFiltering) {
+      const elementInTargetId = this.context.findTargetElementId(
+        sourceElement.id
+      );
+      const isInvalid = !Id64.isValid(elementInTargetId);
+      const sourceInTargetElemProps = this.targetDb.elements.tryGetElementProps(
+        sourceElement.id
+      );
+      const sourceElemProps = this.sourceDb.elements.getElementProps(
+        sourceElement.id
+      );
+
+      if (!isInvalid && elementInTargetId !== sourceElement.id) {
+        // Element found with different id
+        throw new Error(
+          `Element id(${sourceElement.id}) cannot be preserved. Found a different mapping(${elementInTargetId}) from source element`
+        );
+      }
+      // if we don't find mapping for source element in target(invalid) but an element with source id exists in target and it is not the same as source element
+      else if (
+        isInvalid &&
+        sourceInTargetElemProps?.federationGuid &&
+        sourceElemProps?.federationGuid !==
```

Comment:
I dont think we need to compare federationGuids here. If `isInvalid` is already true, AND we found an element in the target with the sameId then we know they didnt have the same federationGuid otherwise further up in onExportElement would've found it when it calls `_queryElemIdByFedGuid`

NOTE: The above statement is actually not true because of preserveElementIdsForFiltering if statement at the top.
I wonder if we should scrap that if statement at the top and worry about it here? Then we can just check the targetElementId instead of calling
`const elementInTargetId = this.context.findTargetElementId(
        sourceElement.id
      );`

Reply:
To address your first statement, I found some tests that were failing and when I investigated. Though invalid, the element we found in the target was the same element as the source i.e. it had the same information as the source element. The only thing that was missing was the federation guid(I was not sure why getElementProps did not bring over the fed guid)